Develop caching

silviu-toderita opened the pull request:
-Added caching by integrating DataQueue library
-GPS time can now be used before Cellular time is available, so that cached messages can have correct time. GPS time usually available within a few seconds, whereas Cellular time can take minutes
-Time zone setting added (for console output only)
-Free RAM console output added

lcfyi commented on proto/src/main.cpp:
```diff
@@ -84,6 +85,13 @@ void loop() {
     }
 
+    // If there is valid time pulled from cellular, get time from GPS (if valid)
+    if(Time.now() < 1609459201){
```
Is there any way around having this magic value? I assume this is some UTC time (that looks like it maps to Friday, January 1, 2021 12:00:01 AM?

silviu-toderita replied:
Yes, it looks like there are indeed functions to check the validity of the time properly! Done

lcfyi commented on lib/Telemetry/src/Telemetry.h:
```diff
@@ -4,28 +4,49 @@
 #endif
 
-void publishMessage(String topic, String& msg) {
+DataQueue dataQ;
```
It might be worth moving this into a `cpp` file in order to encapsulate it. Otherwise, every file that imports `Telemetry.h` will have access to this object, which may not be great.

(At the same time, I don't exactly remember whether a new copy will be created per translation unit or not. I _believe_ it does, which means we'll have multiple copies of `dataQ` if we import `Telemetry.h` in multiple files.

silviu-toderita replied:
Shouldn't the define guards protect against that? If not, I should make a new cpp file and possibly a whole Telemetry class to encapsulate functions common to all vehicles.

lcfyi replied:
Sort of. The define guards prevent multiple inclusions of header files in a single translation unit (ie. a single `.c` or `.cpp` file), but the compiler gets confused during the linking phase since a variable may end up defined in multiple places if you put the definition in the header.

silviu-toderita commented:
Been a bit busy last couple weeks, but I've taken your suggestion and moved the DataQueue object into main.cpp for now. Not sure yet if it's worth creating a Telemetry class, it may overcomplicate things. I'd like to get the new members introduced to the codebase and then we can all decide on how to move forward with it together. 

I've also moved all the Telemetry libraries into a single folder for clarity.

lcfyi commented:
> Been a bit busy last couple weeks, but I've taken your suggestion and moved the DataQueue object into main.cpp for now. Not sure yet if it's worth creating a Telemetry class, it may overcomplicate things. I'd like to get the new members introduced to the codebase and then we can all decide on how to move forward with it together.

That's fair, I agree that a telemetry class might overcomplicate things here. Otherwise, LGTM.
